### Q1- What is aliasing?

In [ ]:
a = 4

In [1]:
id(a)

NameError: name 'a' is not defined

In [ ]:
id(4)

In [3]:
# Aliashing

a = 4
b = a

print(id(a))
print(id(b))
print(id(4))


140710243912216
140710243912216
140710243912216


In [4]:
del a

In [5]:
a


NameError: name 'a' is not defined

In [6]:
print(b)

4


In [25]:
print(id(4))
a = 4
b = a
print(id(a))
print(id(b))
a = 10000
b = a
print(id(a))
print(id(b))
print(id(4))
print(id(10000))

140710243912216
140710243912216
140710243912216
150138928
150138928
140710243912216
150138960


In [26]:
print(id(260))
a = 260
b = a
print(id(a))
print(id(b))
a = 10000
b = a
print(id(a))
print(id(b))
print(id(260))
print(id(10000))

150137040
150137648
150137648
150138288
150138288
150134896
150137360


### **Complete Notes on Aliasing in Python**

#### **What is Aliasing?**

In Python, **aliasing** occurs when two or more variables reference the same object in memory. This means that changes made to the object through one variable will also be reflected in the other variable(s) since they all point to the same memory address.

---

#### **Key Concepts**

1. **Mutable vs. Immutable Types**:
   - **Mutable Types**: Objects like lists, dictionaries, sets, etc., are mutable, meaning their contents can be changed. Aliasing is particularly significant with mutable objects because modifications are shared across aliases.
   - **Immutable Types**: Objects like integers, floats, strings, and tuples are immutable, meaning their contents cannot be changed. For immutable types, aliasing doesn't have the same implications since any "change" creates a new object.

2. **Memory Sharing**:
   - Aliasing results in multiple variables pointing to the same memory location. Python uses this approach to optimize memory usage.

---

#### **Examples of Aliasing**

##### Example 1: Aliasing with Mutable Objects
```python
a = [1, 2, 3]
b = a  # b is an alias for a

# Modify the list using b
b.append(4)

print(a)  # Output: [1, 2, 3, 4]
print(b)  # Output: [1, 2, 3, 4]
```
- **Explanation**: Both `a` and `b` reference the same list object. Changes made through `b` are reflected in `a`.

##### Example 2: Aliasing with Immutable Objects
```python
x = 10
y = x  # y is an alias for x

y += 5  # This creates a new object for y

print(x)  # Output: 10
print(y)  # Output: 15
```
- **Explanation**: Since integers are immutable, modifying `y` creates a new integer object. `x` remains unaffected.

---

#### **Checking Memory Addresses**

In Python, the `id()` function returns the memory address of an object. This can be used to verify aliasing.

```python
a = [1, 2, 3]
b = a

print(id(a))  # Example: 140346284396224
print(id(b))  # Same as id(a): 140346284396224
```

---

#### **Avoiding Unintentional Aliasing**

##### Using Copy for Mutable Objects
To avoid aliasing, you can create a copy of the object.

1. **Shallow Copy**:
   - Creates a new object but does not recursively copy nested objects.
   - Use the `copy()` method or the `copy` module.

   ```python
   import copy
   
   a = [1, 2, [3, 4]]
   b = copy.copy(a)  # Shallow copy

   b[0] = 10
   b[2][0] = 99

   print(a)  # Output: [1, 2, [99, 4]]
   print(b)  # Output: [10, 2, [99, 4]]
   ```

2. **Deep Copy**:
   - Creates a new object and recursively copies all nested objects.
   - Use `copy.deepcopy()`.

   ```python
   c = copy.deepcopy(a)

   c[2][0] = 100
   print(a)  # Output: [1, 2, [99, 4]]
   print(c)  # Output: [1, 2, [100, 4]]
   ```

##### Using Slicing for Lists
```python
a = [1, 2, 3]
b = a[:]  # Creates a shallow copy

b.append(4)
print(a)  # Output: [1, 2, 3]
print(b)  # Output: [1, 2, 3, 4]
```

---

#### **Aliasing in Function Arguments**

When passing mutable objects to a function, aliasing occurs unless explicitly avoided.

```python
def modify_list(lst):
    lst.append(4)

a = [1, 2, 3]
modify_list(a)

print(a)  # Output: [1, 2, 3, 4] (modified due to aliasing)
```

---

#### **Aliasing and Garbage Collection**

Aliasing impacts garbage collection. An object is deleted from memory only when there are no references (aliases) to it.

```python
a = [1, 2, 3]
b = a  # Alias
del a

print(b)  # Output: [1, 2, 3] (object is still accessible via b)
```

---

#### **Advantages of Aliasing**
1. **Memory Optimization**: Efficient use of memory as the same object is reused.
2. **Faster Execution**: Avoids the overhead of creating duplicate objects.

---

#### **Disadvantages of Aliasing**
1. **Unintended Side Effects**: Changes made through one alias affect all aliases, leading to unexpected behavior.
2. **Debugging Complexity**: Tracking changes in aliased objects can be challenging.

---

#### **Key Takeaways**
1. Aliasing occurs when multiple variables point to the same object.
2. It has significant implications for mutable objects.
3. To avoid unintended aliasing, use copying techniques.
4. Always check memory addresses using `id()` to verify aliasing.
5. Understand the behavior of functions with mutable and immutable objects to avoid unintended side effects.

By mastering aliasing, you can write more efficient and predictable Python programs!

### Q2- What is garbage collection?

In [ ]:
a = "DSMP aaa"
b = a
c = b

In [28]:
import sys
sys.getrefcount("DSMP aaa") ### DSMP aaa par kitne variable point kar rhe hai
# yee value sys.getrefcount deta hai  

3

### **Complete Notes on Garbage Collection in Python**

Garbage collection in Python is an automated process that manages memory by reclaiming unused objects to free up space, ensuring efficient use of system resources. It is an essential part of Python’s memory management and helps avoid memory leaks.

---

### **Key Concepts of Garbage Collection**

#### **1. What is Garbage Collection?**
- Garbage collection refers to the process of identifying and deallocating objects that are no longer in use (unreachable) by the program.
- Python’s **garbage collector** is part of its memory management system and works in conjunction with Python’s dynamic memory allocation.

#### **2. Why Garbage Collection is Necessary**
- Prevents memory leaks by automatically freeing up memory occupied by unused objects.
- Ensures efficient memory usage, particularly in long-running programs.
- Simplifies programming as developers don't need to explicitly deallocate memory.

---

### **How Python Manages Memory**

1. **Reference Counting**:
   - Python tracks the number of references to each object using a counter.
   - When an object’s reference count drops to zero, it is considered unreachable and eligible for garbage collection.

   Example:
   ```python
   a = [1, 2, 3]  # Object is created, reference count = 1
   b = a           # Reference count = 2
   del a           # Reference count = 1
   del b           # Reference count = 0, object is garbage collected
   ```

2. **Garbage Collector**:
   - Python’s `gc` (garbage collector) module handles the reclamation of unused memory.
   - It is primarily responsible for dealing with **cyclic references**, where objects reference each other, making them unreachable.

---

### **Garbage Collection Mechanism**

#### **1. Reference Counting**
   - Every object in Python has a reference count that tracks the number of references to it.
   - Reference count increases when:
     - A new variable is assigned to the object.
     - The object is passed as an argument to a function.
     - The object is added to a container like a list or dictionary.
   - Reference count decreases when:
     - A variable referencing the object is deleted.
     - The variable is reassigned to another object.
     - The object is removed from a container.

#### **2. Cyclic Garbage Collection**
   - The garbage collector handles **reference cycles**, which occur when two or more objects reference each other, preventing their reference counts from reaching zero.

   Example of a reference cycle:
   ```python
   class Node:
       def __init__(self, value):
           self.value = value
           self.next = None

   a = Node(1)
   b = Node(2)

   a.next = b
   b.next = a  # Creates a cycle

   del a
   del b
   ```

#### **3. Generational Garbage Collection**
   - Python divides objects into three generations:
     - **Generation 0**: Newly created objects.
     - **Generation 1**: Objects that survived garbage collection in Generation 0.
     - **Generation 2**: Long-lived objects.
   - The garbage collector runs more frequently on younger generations since they are more likely to become unreachable.

---

### **Python `gc` Module**

Python provides the `gc` module to interact with the garbage collection system.

#### **Key Functions**
1. **`gc.collect()`**:
   - Manually triggers garbage collection.
   - Returns the number of unreachable objects collected.

   ```python
   import gc

   gc.collect()
   ```

2. **`gc.isenabled()`**:
   - Checks whether automatic garbage collection is enabled.

   ```python
   print(gc.isenabled())  # Output: True or False
   ```

3. **`gc.disable()`**:
   - Disables automatic garbage collection.

   ```python
   gc.disable()
   ```

4. **`gc.enable()`**:
   - Enables automatic garbage collection.

   ```python
   gc.enable()
   ```

5. **`gc.get_objects()`**:
   - Returns a list of all objects tracked by the garbage collector.

   ```python
   print(gc.get_objects())
   ```

6. **`gc.get_stats()`**:
   - Provides statistics about the garbage collection process.

   ```python
   print(gc.get_stats())
   ```

---

### **Garbage Collection in Action**

#### **Example 1: Reference Count**
```python
import sys

a = [1, 2, 3]
print(sys.getrefcount(a))  # Reference count of `a`

b = a  # Reference count increases
print(sys.getrefcount(a))

del b  # Reference count decreases
print(sys.getrefcount(a))
```

#### **Example 2: Cyclic References**
```python
import gc

class Node:
    def __init__(self, value):
        self.value = value
        self.next = None

a = Node(1)
b = Node(2)

a.next = b
b.next = a  # Creates a cycle

del a
del b

# Manually run garbage collection
gc.collect()
```

---

### **Advantages of Python’s Garbage Collection**
1. **Automatic Memory Management**:
   - Reduces the burden on developers to manage memory manually.
2. **Cyclic Reference Handling**:
   - Automatically identifies and cleans up reference cycles.
3. **Efficient Memory Usage**:
   - Optimizes memory usage through generational collection.

---

### **Limitations of Garbage Collection**
1. **Performance Overhead**:
   - Garbage collection adds processing overhead, especially for large programs.
2. **Non-Deterministic**:
   - Objects are not guaranteed to be garbage collected immediately after they become unreachable.
3. **Manual Intervention**:
   - In some cases, manual garbage collection may be necessary to reclaim memory promptly.

---

### **Best Practices to Avoid Memory Leaks**

1. **Avoid Cyclic References**:
   - Use weak references (`weakref` module) to break reference cycles.
   
   Example:
   ```python
   import weakref

   class Node:
       def __init__(self, value):
           self.value = value
           self.next = None

   a = Node(1)
   b = Node(2)

   a.next = weakref.ref(b)
   b.next = weakref.ref(a)
   ```

2. **Release Unnecessary References**:
   - Use `del` to remove references when they are no longer needed.

3. **Minimize Global Variables**:
   - Global variables increase the lifespan of objects, making them less likely to be garbage collected.

4. **Profile Memory Usage**:
   - Use tools like `tracemalloc` to track memory usage and identify leaks.

---

### **Conclusion**

Garbage collection in Python is an efficient and automatic mechanism to manage memory, reclaim unused objects, and handle cyclic references. Understanding its behavior and limitations allows developers to write efficient, memory-optimized programs while avoiding common pitfalls like memory leaks.

### 3- What is mutability and why is it dangerous in certain scenarios?

In [29]:
L = [1,2,3,4]
print(id(L))

L[0] = 6
print(id(L))

8421824
8421824


In [40]:
T = (1,2,3)
print(id(T))

T = T + (4,)
print(T)
print(id(T))

8325760
(1, 2, 3, 4)
150163984


In [41]:
# Mutability problem 1

a = [1,2,3]
b = a

b.append(4)
print(b)
print(a)

[1, 2, 3, 4]
[1, 2, 3, 4]


In [42]:
# Mutability problem 2

def func(data):
    data.append(4)
    
a = [1,2,3]
func(a)
print(a)

[1, 2, 3, 4]


In [43]:
def func(data):
    data = data + (4,)
    
a = (1,2,3)
func(a)
print(a)

(1, 2, 3)


### **Complete Notes on Mutability in Python**

---

### **What is Mutability in Python?**

Mutability refers to the ability of an object to be modified after its creation. In Python, objects are categorized as **mutable** or **immutable** based on whether their state (data) can change after they are created.

---

### **Mutable vs Immutable Objects**

| **Aspect**         | **Mutable Objects**                       | **Immutable Objects**                |
|---------------------|-------------------------------------------|---------------------------------------|
| **Definition**      | Objects whose state can be modified.     | Objects whose state cannot be changed.|
| **Examples**        | Lists, Dictionaries, Sets, User-defined classes. | Tuples, Strings, Integers, Floats, Booleans. |
| **Modification**    | Elements can be changed, added, or removed. | Modification creates a new object.    |
| **Performance**     | More memory-efficient when updates are frequent. | Safer and simpler due to immutability. |

---

### **Examples of Mutable and Immutable Objects**

#### **Mutable Example**
```python
# Mutable: List
my_list = [1, 2, 3]
my_list[0] = 10  # Modify the first element
print(my_list)   # Output: [10, 2, 3]
```

#### **Immutable Example**
```python
# Immutable: Tuple
my_tuple = (1, 2, 3)
# my_tuple[0] = 10  # This would raise a TypeError
new_tuple = (10,) + my_tuple[1:]  # Create a new tuple
print(new_tuple)  # Output: (10, 2, 3)
```

---

### **Why is Mutability Dangerous in Certain Scenarios?**

Mutability, while useful, can lead to unintended consequences and bugs if not carefully managed. Below are specific dangers associated with mutable objects:

---

#### **1. Shared References**
Mutable objects can have multiple references pointing to the same memory location. Modifying the object via one reference affects all references, leading to unintended side effects.

**Example: Shared References**
```python
list1 = [1, 2, 3]
list2 = list1  # Both point to the same list
list2[0] = 10  # Modify list2
print(list1)   # Output: [10, 2, 3] (list1 is also affected)
```

**Solution: Use Copies**
```python
list1 = [1, 2, 3]
list2 = list1.copy()  # Create a copy
list2[0] = 10
print(list1)  # Output: [1, 2, 3] (list1 remains unchanged)
```

---

#### **2. Unexpected Behavior in Default Mutable Arguments**
Using mutable objects as default arguments in functions can lead to unexpected results because the default value is shared across all calls to the function.

**Example: Mutable Default Argument**
```python
def add_item(item, my_list=[]):
    my_list.append(item)
    return my_list

print(add_item(1))  # Output: [1]
print(add_item(2))  # Output: [1, 2] (unexpected behavior)
```

**Solution: Use `None` as Default**
```python
def add_item(item, my_list=None):
    if my_list is None:
        my_list = []  # Create a new list
    my_list.append(item)
    return my_list

print(add_item(1))  # Output: [1]
print(add_item(2))  # Output: [2]
```

---

#### **3. Hashability Issues with Mutable Objects**
Mutable objects cannot be used as keys in dictionaries or elements in sets because their hash value can change.

**Example: Hashability**
```python
my_dict = {}
my_list = [1, 2, 3]
# my_dict[my_list] = "value"  # Raises TypeError: unhashable type: 'list'
```

**Solution: Use Immutable Types**
```python
my_dict = {}
my_tuple = (1, 2, 3)
my_dict[my_tuple] = "value"  # Works because tuples are immutable
```

---

#### **4. Debugging Challenges**
Changes to mutable objects can be difficult to trace, especially in large programs where objects are passed between multiple functions or classes.

**Example: Debugging Complexity**
```python
def modify_list(lst):
    lst.append(10)
    return lst

my_list = [1, 2, 3]
result = modify_list(my_list)
print(my_list)  # Output: [1, 2, 3, 10] (modified in place)
```

**Solution: Avoid Modifying Objects in Place**
```python
def modify_list(lst):
    new_list = lst + [10]  # Return a new list
    return new_list

my_list = [1, 2, 3]
result = modify_list(my_list)
print(my_list)  # Output: [1, 2, 3] (original list remains unchanged)
```

---

### **Benefits of Immutability**

1. **Predictable Behavior**:
   - Immutable objects prevent unintended modifications.
2. **Hashable and Efficient**:
   - Immutable objects can be used as keys in dictionaries and elements in sets.
3. **Thread-Safe**:
   - Immutable objects are inherently safe for concurrent access.

---

### **Managing Mutability in Python**

#### **1. Use Immutable Types**
- Favor tuples over lists when immutability is required.
- Use frozensets instead of sets for immutable collections.

#### **2. Create Copies**
- When working with mutable objects, create copies to avoid modifying the original.

#### **3. Avoid In-Place Modifications**
- Prefer creating new objects over modifying existing ones.

#### **4. Use Data Classes with `frozen=True`**
- Python’s `dataclasses` module allows you to create immutable classes using the `frozen` parameter.

**Example: Immutable Data Class**
```python
from dataclasses import dataclass

@dataclass(frozen=True)
class Point:
    x: int
    y: int

p = Point(1, 2)
# p.x = 10  # Raises FrozenInstanceError
```

---

### **Key Takeaways**

- **Mutability** is a core feature of Python objects that allows modification after creation.
- While useful, it can lead to **unexpected behavior** if not carefully managed, especially with shared references, default arguments, and hashability.
- Favor **immutable objects** in scenarios where predictability, safety, and efficiency are critical.
- Use techniques like creating copies, avoiding in-place modifications, and leveraging immutable data structures to mitigate risks associated with mutability.

By understanding and managing mutability effectively, Python developers can write more robust, maintainable, and error-free code.

### 4- What is cloning?

In [45]:
a = [1,2,3]
# Cloning
b = a[:]

print(id(a))
print(id(b))

b.append(4)
print(b)
print(a)


8353664
8397568
[1, 2, 3, 4]
[1, 2, 3]


In [46]:
def func(data):
    data.append(4)
    
a = [1,2,3]
func(a[:])
print(a)

[1, 2, 3]


### **Cloning in Python**

Cloning is the process of creating a duplicate object that has the same content as the original but resides in a different memory location. In Python, cloning is often used for mutable objects to ensure changes to the clone do not affect the original object.

---

### **Cloning vs Assignment**

- **Assignment (`=`)**: Creates a new reference to the same object. Both references point to the same memory location, so changes in one reflect in the other.
- **Cloning**: Creates a new object with the same data as the original but with a different memory location.

---

### **Cloning a List Using Slicing**

In Python, lists can be cloned using slicing. The slicing syntax `a[:]` creates a shallow copy of the list `a`.

#### **Code Example**
```python
a = [1, 2, 3]

# Cloning using slicing
b = a[:]

print("ID of a:", id(a))  # Memory location of 'a'
print("ID of b:", id(b))  # Memory location of 'b', different from 'a'

# Modifying the clone
b.append(4)
print("Cloned list (b):", b)  # Output: [1, 2, 3, 4]
print("Original list (a):", a)  # Output: [1, 2, 3] (unchanged)
```

---

### **Key Observations**

1. **Different Memory Locations**:
   - `id(a)` and `id(b)` show that `a` and `b` are stored in different memory locations.
   - This confirms that `b` is a clone, not a reference to the same object.

2. **Independence of Objects**:
   - Changes made to `b` do not affect `a`.

---

### **Types of Cloning**

#### **1. Shallow Cloning**
- A shallow copy creates a new object but only copies references to the original elements (for nested objects).
- **Example: Slicing**
  ```python
  a = [[1, 2], [3, 4]]
  b = a[:]
  b[0][0] = 99
  print(a)  # Output: [[99, 2], [3, 4]] (inner lists are still shared)
  ```

#### **2. Deep Cloning**
- A deep copy creates a completely independent copy, including all nested objects.
- Requires the `copy` module.
  ```python
  import copy

  a = [[1, 2], [3, 4]]
  b = copy.deepcopy(a)
  b[0][0] = 99
  print(a)  # Output: [[1, 2], [3, 4]] (no shared references)
  ```

---

### **Cloning Techniques in Python**

1. **Using Slicing**:
   - `b = a[:]`
   - Works only for lists and creates a shallow copy.

2. **Using `copy.copy()`**:
   - `import copy; b = copy.copy(a)`
   - Creates a shallow copy and works for most objects.

3. **Using `copy.deepcopy()`**:
   - `import copy; b = copy.deepcopy(a)`
   - Creates a deep copy and ensures independence of nested objects.

4. **Using List Comprehension**:
   - `b = [item for item in a]`
   - Similar to slicing, creates a shallow copy.

5. **Using the `list()` Constructor**:
   - `b = list(a)`
   - Creates a shallow copy of the list.

---

### **When to Use Cloning**

- **Avoid Side Effects**:
  - To prevent changes in one object from affecting another.
- **Work with Mutable Objects**:
  - Especially for lists, dictionaries, and user-defined mutable objects.
- **Handle Nested Structures**:
  - Use deep cloning for nested objects to ensure complete independence.

---

### **Advantages of Cloning**

- **Ensures Independence**:
  - Modifications to the clone don’t affect the original.
- **Control over Data**:
  - Useful when passing objects to functions or working in multi-threaded environments.

---

By understanding cloning and its nuances, you can manage mutable objects more effectively in Python, ensuring robust and error-free code.

### 5- Difference between shallow copy and deep copy ?

In [48]:
a = [1,2,3]
# shallow copy 
b = a.copy()

b.append(4)
print(a)
print(b)

[1, 2, 3]
[1, 2, 3, 4]


In [50]:
# shallow copy problem

a = [1,2,3,[4,5]]
b = a.copy()
print(b)

b[-1][0] = 400
print(b)
print(a)

print(id(a))
print(id(b))

print(id(a[-1][0]))
print(id(b[-1][0]))

[1, 2, 3, [4, 5]]
[1, 2, 3, [400, 5]]
[1, 2, 3, [400, 5]]
8343424
8347648
150138928
150138928


In [52]:
import copy
c = copy.deepcopy(a)

c[-1][0] = 1000
print(c)
print(a)

print(id(a[-1][0]))
print(id(c[-1][0]))

[1, 2, 3, [1000, 5]]
[1, 2, 3, [400, 5]]
150138928
150126768


### **Deep vs Shallow Copies in Python**

Copying in Python allows the creation of new objects with data from an existing object. However, there are **two types of copies**: **shallow copy** and **deep copy**, and the choice between them depends on how the objects and their nested structures are handled.

---

### **Shallow Copy**
- **Definition**: A shallow copy creates a new object but **does not create copies of the objects nested inside** (if the original object contains other objects, they are still shared).
- Only the outermost object is copied, and inner objects retain references to the same memory locations as the original.

---

### **Deep Copy**
- **Definition**: A deep copy creates a completely new object **and recursively copies all nested objects** inside it, ensuring no shared references between the original and copied objects.

---

### **Key Differences Between Shallow and Deep Copy**

| **Aspect**            | **Shallow Copy**                              | **Deep Copy**                                  |
|-----------------------|-----------------------------------------------|-----------------------------------------------|
| **What is copied?**    | Only the outer object is copied.              | The outer object and all nested objects are copied. |
| **Nested Objects**     | References of nested objects are shared.      | Nested objects are independently copied.      |
| **Dependency**         | Changes to nested objects in the copy affect the original. | No dependency; changes do not affect each other. |
| **Performance**        | Faster but less secure for complex objects.   | Slower but ensures full independence.         |
| **Use Case**           | Suitable for flat or non-nested objects.      | Required for nested or hierarchical objects.  |

---

### **Implementation**

#### **Shallow Copy**
1. Using slicing: `b = a[:]`
2. Using `copy.copy()`: 
   ```python
   import copy
   b = copy.copy(a)
   ```

#### **Deep Copy**
1. Using `copy.deepcopy()`:
   ```python
   import copy
   b = copy.deepcopy(a)
   ```

---

### **Memory-Level Example**

Let’s use a **nested list** example to demonstrate shallow and deep copies.

#### **Code Example**
```python
import copy

# Original nested list
original = [[1, 2, 3], [4, 5, 6]]

# Shallow Copy
shallow = copy.copy(original)

# Deep Copy
deep = copy.deepcopy(original)

# Memory locations
print(f"Original: {id(original)}")              # Memory location of the outer object
print(f"Shallow: {id(shallow)}")                # Different from `original`
print(f"Deep: {id(deep)}")                      # Different from `original`

print(f"Original Inner: {id(original[0])}")     # Inner list's memory location
print(f"Shallow Inner: {id(shallow[0])}")       # Same as `original`
print(f"Deep Inner: {id(deep[0])}")             # Different from `original`

# Modify shallow copy
shallow[0][0] = 99
print(f"Original after shallow modification: {original}")  # Affected

# Modify deep copy
deep[0][0] = 88
print(f"Original after deep modification: {original}")     # Not affected
```

---

### **Explanation**

1. **Memory Allocation**:
   - For a **shallow copy**, the outer object is new (different `id`), but references to inner objects are the same.
   - For a **deep copy**, both the outer object and all inner objects have new memory locations.

2. **Effect of Changes**:
   - **Shallow Copy**: When modifying an inner object, the change is reflected in the original because they share the same reference.
   - **Deep Copy**: Modifying the deep copy does not affect the original because inner objects are independently copied.

#### **Output**
```plaintext
Original: 139965279083904
Shallow: 139965279085632
Deep: 139965279086032
Original Inner: 139965279084224
Shallow Inner: 139965279084224
Deep Inner: 139965279085184
Original after shallow modification: [[99, 2, 3], [4, 5, 6]]
Original after deep modification: [[99, 2, 3], [4, 5, 6]]
```

---

### **Shallow Copy: Benefits and Risks**
- **Benefits**:
  - Faster to create as it doesn’t copy nested objects.
  - Memory-efficient for flat structures.

- **Risks**:
  - Modifications in nested structures affect the original, leading to unexpected side effects.

---

### **Deep Copy: Benefits and Risks**
- **Benefits**:
  - Ensures full independence between the original and the copied object.
  - Safe for nested structures, making it ideal for hierarchical data.

- **Risks**:
  - Slower and more memory-intensive due to recursive copying.
  - May not work for custom objects with non-copyable attributes.

---

### **When to Use**
- Use **shallow copy** for flat objects or when inner objects do not need to be independent.
- Use **deep copy** when working with nested objects that must be fully decoupled.

---

### **Comparison with Real-Life Example**

- **Shallow Copy**: Imagine duplicating a folder structure where the outer folder is new, but all files inside are still linked to the original folder.
- **Deep Copy**: Duplicating a folder structure where both the folder and all its contents are completely independent copies.

---

Understanding the distinction between shallow and deep copies in Python is crucial for managing object dependencies effectively and avoiding unintended side effects in your programs.

### Q6- How nested lists are stored in memory?

### **How Nested Lists Are Stored in Memory in Python**

A **nested list** in Python is a list that contains other lists as its elements. Understanding how Python stores nested lists in memory helps developers write efficient code and avoid unintended side effects caused by reference sharing.

---

### **Key Concepts of Nested List Storage**
1. **Lists Are Objects**:
   - In Python, lists are objects stored in memory.
   - A list contains references (memory addresses) to its elements, rather than the actual data.

2. **Nested Lists Store References**:
   - For nested lists, the outer list holds references to the inner lists.
   - The inner lists themselves are independent objects in memory with their unique memory addresses.

3. **Mutability and References**:
   - Lists in Python are mutable. Any modification to an inner list via a reference affects all variables pointing to that list.

4. **Shared References**:
   - If multiple variables refer to the same inner list, changes made through one variable will be visible in others.

---

### **Memory Representation of Nested Lists**

#### Example
```python
nested_list = [[1, 2, 3], [4, 5, 6], [7, 8, 9]]
```

1. The **outer list**:
   - A single object in memory that contains **references** (pointers) to three inner lists.

2. The **inner lists**:
   - Each inner list is a separate object in memory, holding references to its elements.

3. The **elements**:
   - Each element (e.g., `1`, `2`, etc.) is stored independently in memory.

---

#### **Memory Diagram**
Assuming `nested_list = [[1, 2, 3], [4, 5, 6], [7, 8, 9]]`:
- `nested_list` (outer list) is stored at `0x1a2b3c`.
- Each inner list (e.g., `[1, 2, 3]`) has its own memory address:
  - `[1, 2, 3]` → `0x4d5e6f`
  - `[4, 5, 6]` → `0x7g8h9i`
  - `[7, 8, 9]` → `0x1j2k3l`
- Each integer (e.g., `1`, `2`, etc.) is stored at a separate location.

---

### **Example Code and Explanation**

#### Code
```python
nested_list = [[1, 2, 3], [4, 5, 6], [7, 8, 9]]

# Outer list memory address
print(f"Outer list ID: {id(nested_list)}")

# Inner list memory addresses
for i, sublist in enumerate(nested_list):
    print(f"Inner list {i} ID: {id(sublist)}")

# Element memory addresses
for i, sublist in enumerate(nested_list):
    for j, element in enumerate(sublist):
        print(f"Element {element} ID: {id(element)}")
```

#### Output
```plaintext
Outer list ID: 140702872837328
Inner list 0 ID: 140702872837184
Inner list 1 ID: 140702872837248
Inner list 2 ID: 140702872837312
Element 1 ID: 9788928
Element 2 ID: 9788960
Element 3 ID: 9788992
...
```

---

### **Modifying Nested Lists**

#### Example 1: Modifying Inner Lists
```python
nested_list = [[1, 2], [3, 4]]

# Modify an inner list
nested_list[0][0] = 99
print(nested_list)  # Output: [[99, 2], [3, 4]]
```
Explanation:
- The outer list reference is unchanged.
- The first inner list is updated as it shares the same memory address.

#### Example 2: Adding New Inner Lists
```python
nested_list.append([5, 6])
print(nested_list)  # Output: [[1, 2], [3, 4], [5, 6]]
```
Explanation:
- A new list is created in memory, and its reference is added to the outer list.

#### Example 3: Shared References
```python
shared = [1, 2]
nested_list = [shared, shared]

# Modify the shared list
shared[0] = 99
print(nested_list)  # Output: [[99, 2], [99, 2]]
```
Explanation:
- Both references in the outer list point to the same inner list. Modifying one affects all.

---

### **Copying Nested Lists**

#### **Shallow Copy**
```python
import copy

nested_list = [[1, 2], [3, 4]]
shallow_copy = copy.copy(nested_list)

# Modify inner list in shallow copy
shallow_copy[0][0] = 99
print(nested_list)  # Output: [[99, 2], [3, 4]]
```
Explanation:
- The outer list is copied, but inner lists are shared. Changes in the copy affect the original.

#### **Deep Copy**
```python
deep_copy = copy.deepcopy(nested_list)

# Modify inner list in deep copy
deep_copy[0][0] = 99
print(nested_list)  # Output: [[1, 2], [3, 4]]
```
Explanation:
- Both outer and inner lists are independently copied. Changes in the copy do not affect the original.

---

### **Best Practices with Nested Lists**

1. **Avoid Shared References**:
   - Be cautious when appending or assigning inner lists to avoid unintended shared references.

2. **Use `copy.deepcopy()`**:
   - When working with nested lists that require independent copies, always use `deepcopy`.

3. **Mutability Awareness**:
   - Remember that inner lists are mutable, and changes through one reference affect all shared references.

4. **Use Iterators Carefully**:
   - Modifications during iteration may cause unexpected behavior. Prefer copying the list before iterating.

---

### **Visualizing Nested Lists in Memory**

1. **Original Nested List**:
   ```
   Outer List: [ [List 1], [List 2], [List 3] ]
   List 1: [1, 2, 3]
   List 2: [4, 5, 6]
   List 3: [7, 8, 9]
   ```

2. **Memory References**:
   ```
   Outer List → 0xABC
   List 1 → 0xDEF → [1, 2, 3]
   List 2 → 0xGHI → [4, 5, 6]
   List 3 → 0xJKL → [7, 8, 9]
   ```

---

### **Conclusion**

- Nested lists in Python store references to inner lists, and these lists can independently reference other objects.
- Understanding reference behavior is crucial for managing nested lists effectively and avoiding unintended side effects.
- Choose between shallow and deep copies based on whether you want to share or duplicate inner list data.

### Q7- How strings are stored in memory

In [37]:
s = "Man is Good"

print(id(s))
print(id(s[0]))
print(id("M"))

print(id(s[1]))
print(id('a'))

print(id(s[2]))
print(id("n"))

81441840
140715532270872
140715532270872
140715532271832
140715532242144
140715532272456
140715532256712


The behavior you are observing is due to Python's **string interning** and **optimization mechanisms**. Let's break down the logic behind why some letters in the string `s` share the same memory address while others do not:

---

### **1. String Interning**
- **String interning** is a mechanism where Python tries to reuse immutable objects (like strings) to save memory and improve performance.
- Interned strings include:
  - Strings that are valid Python identifiers (e.g., `"abc"`, `"xyz"`).
  - Short strings (usually ≤ 20 characters, but it depends on implementation).
  - Literal single characters like `'a'`, `'b'`, etc., if frequently used.

---

### **2. Immutable Nature of Strings**
- Strings in Python are **immutable**. When you access a character like `s[0]` or `s[1]`, Python does not create a new string object. Instead, it retrieves a reference to the existing string or character in memory.
- If the character is already interned, its memory address will match another instance of the same character.

---

### **3. Why Some Letters Share the Same Address?**
- For the given example:
  ```python
  s = "Man is Good"
  ```
  - `s[0]` → `'M'` (a single uppercase letter, which is likely interned).
  - `"M"` → Same as `s[0]`, as `'M'` is interned.
  - `s[1]` → `'a'` (a single lowercase letter, also interned, but at a different address).
  - `"a"` → Matches `s[1]` due to interning.
  - `s[2]` → `'n'` (lowercase letter `'n'` is also interned, but may or may not share the address due to how Python optimizes literals).

---

### **4. Why Some Letters Don't Share the Same Address?**
- **Temporary Character Objects**:
  - When you access a specific character via indexing (e.g., `s[2]`), Python may create a temporary object for that character if it is not already interned or if the optimization does not apply.
  - Example:
    - `s[2]` gives `'n'`, which may not match `id("n")` if the `"n"` string is already interned while `s[2]` was created temporarily.

---

### **5. Factors Influencing Memory Addresses**
- **Interning Rules**:
  - Single characters (like `'a'`, `'M'`) are more likely to be interned.
  - Strings containing spaces or combinations of characters are not always interned.
- **Context of Creation**:
  - If a character is created in a specific context (e.g., indexing into a string), it may be a temporary object.
- **Implementation Details**:
  - Interning behavior depends on the Python version and implementation (e.g., CPython).

---

### **Memory-Level Explanation**

| Operation | Memory Address | Reasoning |
|-----------|----------------|-----------|
| `id(s)`   | `81441840`     | Memory address of the entire string object `"Man is Good"`. |
| `id(s[0])` | `140715532270872` | `'M'` is interned as a single character. |
| `id("M")` | `140715532270872` | Interned single-character string, same as `s[0]`. |
| `id(s[1])` | `140715532271832` | `'a'` is interned but may come from a different context. |
| `id('a')` | `140715532242144` | Interned single-character string, matches other `'a'` references in the same context. |
| `id(s[2])` | `140715532272456` | `'n'` is interned, but the address may differ if created as a temporary object. |
| `id("n")` | `140715532256712` | Interned string `"n"`, a separate object in memory. |

---

### **Key Takeaways**
1. **String interning** applies to frequently used and single-character strings to optimize memory usage.
2. Characters accessed via indexing (`s[i]`) may or may not refer to the same memory location as literals due to temporary object creation.
3. Python's memory model ensures immutability and reusability but has specific rules for interning and object creation.

---

### **How to Check Interning Behavior**
- Use the `sys.intern()` function to force interning and compare addresses explicitly:
  ```python
  import sys
  a = sys.intern("M")
  b = "M"
  print(id(a) == id(b))  # True
  ```

This ensures consistent behavior for string memory management!  

### Further Experimentation

Example 1: Interned Characters
```python
a = 'a'
b = 'a'

print(id(a) == id(b))  # True: 'a' is interned
```
Example 2: Space Not Interned
```python
space1 = ' '
space2 = ' '

print(id(space1) == id(space2))  # False: spaces are dynamically allocated
```

Example 3: Dynamic Allocation for Slices
```python
s = "Man is Good"
n1 = s[0]  # 'M'
n2 = s[4]  # ' '
```
print(id(n1) == id("M"))  # True: 'M' is interned
print(id(n2) == id(" "))  # False: ' ' is not interned

### **How Strings Are Stored in Memory in Python**

Python's handling of strings is both efficient and flexible, which makes it a popular choice for text manipulation. Strings are **immutable** sequences of Unicode characters, and understanding their memory representation is essential for efficient programming.

---

### **Key Characteristics of Strings in Python**

1. **Immutability**:
   - Once created, strings cannot be modified. Any operation that appears to modify a string actually creates a new string object.

2. **Unicode Representation**:
   - Strings in Python are sequences of Unicode characters, allowing representation of international characters and symbols.

3. **Memory Efficiency**:
   - Python uses **string interning** to save memory and improve performance for frequently used strings.

---

### **Memory Representation of Strings**

#### **1. Memory Allocation**
- Strings are stored as objects in memory.
- Each string object has:
  - **Metadata**: Information such as the string length.
  - **Pointer**: Reference to the actual character data.
  - **Character Data**: The Unicode characters stored in the string.

#### **2. String Interning**
- Python interns certain strings to save memory.
  - **Interned Strings**: Short strings or strings that look like identifiers (e.g., `"hello"`, `"abc123"`) are stored in a shared memory pool.
  - Benefits:
    - Reduces memory usage for frequently used strings.
    - Speeds up string comparison because interned strings can be compared by memory addresses instead of content.

#### **3. Dynamic Memory Allocation**
- Non-interned strings are stored in the heap memory.
- A new memory block is allocated whenever a new string is created (except for interned strings).

---

### **String Interning Example**

```python
a = "hello"
b = "hello"

print(id(a))  # Memory address of 'a'
print(id(b))  # Memory address of 'b'

print(a is b)  # True because of string interning
```

#### Output:
```plaintext
140339487687728
140339487687728
True
```

---

### **Strings as Immutable Objects**

1. **Behavior**
   - Modifying a string creates a new string object instead of altering the original.

2. **Example**
   ```python
   s = "hello"
   print(id(s))  # Memory address of the string 'hello'

   s += " world"
   print(id(s))  # Memory address changes because a new string object is created
   ```

   #### Output:
   ```plaintext
   140339487687728
   140339487688304
   ```

---

### **Memory Optimization Techniques**

1. **String Interning**:
   - Python automatically interns:
     - Short strings (e.g., `x = "a"`, `y = "a"`) to save memory.
     - Strings that are valid identifiers (e.g., variable names).

2. **Explicit Interning**:
   - Use the `sys.intern()` function for manual interning.
   ```python
   import sys

   a = sys.intern("this_is_a_long_string")
   b = sys.intern("this_is_a_long_string")

   print(a is b)  # True
   ```

3. **Garbage Collection**:
   - Unused strings are garbage-collected if no references exist.

---

### **String Copy Behavior**

#### **1. Strings Are Shared**
- When assigning a string to a new variable, Python does not copy the string; it creates a new reference to the same object.

```python
a = "hello"
b = a  # Both point to the same memory address
print(id(a) == id(b))  # True
```

#### **2. New Object on Modification**
- Any modification creates a new string object.
```python
a = "hello"
b = a
a += " world"

print(a)  # hello world
print(b)  # hello
print(id(a) == id(b))  # False
```

---

### **String Encoding in Memory**

1. **Unicode Storage**:
   - Strings are stored as Unicode by default.
   - Python uses UTF-8 encoding internally for Unicode support.

2. **Bytes Representation**:
   - Strings can be encoded into bytes for compact storage or transmission.
   ```python
   s = "hello"
   b = s.encode("utf-8")
   print(b)  # b'hello'
   ```

---

### **Memory-Level Explanation**

#### **Step-by-Step Memory Representation**
```python
s = "hello"
```
1. Python checks if `"hello"` is in the interned pool.
   - If yes: Uses the existing memory location.
   - If no: Allocates memory, interns `"hello"`, and returns its reference.

2. The variable `s` stores a reference (pointer) to the string object.

---

### **Efficient String Manipulation**

1. **Avoid Repeated Concatenation**:
   - Repeated concatenation creates multiple intermediate strings.
   - Use `str.join()` for efficient concatenation.
   ```python
   words = ["Python", "is", "great"]
   result = " ".join(words)
   print(result)  # Python is great
   ```

2. **Use Immutable Strings Wisely**:
   - Strings are immutable, so modifications create new objects.
   - Minimize modifications to avoid memory overhead.

---

### **Common Issues with Strings in Memory**

1. **Unintended Sharing**:
   - Variables referencing the same interned string can cause confusion.
   - Example:
     ```python
     a = "hello"
     b = "hello"
     a = a + " world"
     print(b)  # Still "hello"
     ```

2. **Memory Overhead**:
   - Strings can consume significant memory in large applications.
   - Use bytes or memory views for compact storage if strings are large.

---

### **Conclusion**

- Strings in Python are immutable and stored as Unicode objects in memory.
- String interning optimizes memory usage for common strings.
- Python's memory model for strings ensures efficiency and flexibility, but understanding the behavior of references and immutability is crucial to avoid pitfalls.
- Use best practices like interning, efficient concatenation, and encoding to manage string memory effectively.

### Q8 - Why tuples take less memory than lists?

Tuples take less memory than lists in Python because of their **immutability** and simpler data structure. Here’s a detailed explanation:

---

### 1. **Immutability Reduces Overhead**
   - **Lists are mutable**: Python must allocate extra memory and include additional metadata to handle operations like resizing, appending, or deleting elements in a list.
   - **Tuples are immutable**: Since tuples cannot be changed after creation, Python can allocate memory more efficiently without the need to maintain overhead for mutability.

   **Example**:
   - A list requires pointers for each element and extra memory for dynamic resizing.
   - A tuple only needs fixed pointers for its elements, saving memory.

---

### 2. **No Extra Buffer for Growth**
   - Lists in Python use a **dynamic array** under the hood. To optimize performance during resizing, Python preallocates extra memory for lists (usually with exponential growth).
   - Tuples, being immutable, do not require this extra buffer. Python allocates exactly the amount of memory needed for the tuple elements.

   **Result**: Tuples have a smaller memory footprint.

---

### 3. **Memory Layout**
   - **Lists**: Store metadata for mutability, including the current size, allocated capacity, and pointers to the elements.
   - **Tuples**: Store only the fixed number of pointers corresponding to the elements.

   **Memory usage example**:
   ```python
   import sys

   list_example = [1, 2, 3]
   tuple_example = (1, 2, 3)

   print(sys.getsizeof(list_example))  # Memory used by the list
   print(sys.getsizeof(tuple_example))  # Memory used by the tuple
   ```

   **Typical Output**:
   ```
   88  # Memory in bytes for the list
   72  # Memory in bytes for the tuple
   ```

---

### 4. **Simpler Internal Structure**
   - **Lists** include methods and mechanisms for operations like `append()`, `remove()`, `extend()`, etc.
   - **Tuples**, being immutable, lack such methods, resulting in a simpler and smaller internal structure.

---

### 5. **Cache Optimization**
   - Since tuples are immutable, Python can optimize their storage by reusing memory for identical tuples (interning).
   - Lists cannot benefit from such optimizations because their contents can change at any time.

---

### Practical Example of Memory Difference

```python
import sys

# Comparing memory usage of lists and tuples with the same content
list1 = [1, 2, 3, 4, 5]
tuple1 = (1, 2, 3, 4, 5)

print("List size:", sys.getsizeof(list1))   # Example: 96 bytes
print("Tuple size:", sys.getsizeof(tuple1)) # Example: 80 bytes
```

**Why?**
- The list requires extra memory for dynamic resizing and metadata.
- The tuple uses memory only for storing fixed pointers to its elements.

---

### Summary: Why Tuples Use Less Memory
| Feature               | Tuple                          | List                          |
|-----------------------|--------------------------------|-------------------------------|
| **Mutability**        | Immutable (no resizing)       | Mutable (extra metadata)      |
| **Resizing Buffer**   | None                          | Yes, for dynamic growth       |
| **Memory Overhead**   | Minimal (fixed size)          | Higher (dynamic size)         |
| **Methods Support**   | Few methods (simpler object)  | Many methods (complex object) |
| **Reuse Optimization**| Interning for identical tuples| Not applicable                |

Thus, **tuples** are more lightweight, faster, and memory-efficient than **lists** in Python.